File: arch/agat/agat.cc
#include "lib/core/globals.h"
#include "lib/decoders/decoders.h"
#include "agat.h"
#include "lib/core/bytes.h"
#include "fmt/format.h"

uint8_t agatChecksum(const Bytes& bytes)
{
    uint16_t checksum = 0;

    for (uint8_t b : bytes)
    {
        if (checksum > 0xff)
            checksum = (checksum + 1) & 0xff;

        checksum += b;
    }

    return checksum & 0xff;
}

File: java/com/cowlark/fluxengine/arch/agat/Agat.java
package com.cowlark.fluxengine.arch.agat;

import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.Bytes;

/**
 * Constants and helpers for the Agat format, ported from
 * arch/agat/agat.h and arch/agat/agat.cc.
 */
public final class Agat
{
    public static final int AGAT_SECTOR_SIZE = 256;

    public static final long SECTOR_ID = 0x8924555549111444L;
    public static final long DATA_ID = 0x8924555514444911L;

    private Agat()
    {
    }

    public static int agatChecksum(Bytes bytes)
    {
        ByteReader br = new ByteReader(bytes);
        int checksum = 0;

        while (!br.eof())
        {
            int b = br.read8();
            if (checksum > 0xff)
                checksum = (checksum + 1) & 0xff;

            checksum += b;
        }

        return checksum & 0xff;
    }
}